torch/csrc/jit/unpickler.cpp

```diff
@@ -456,6 +457,16 @@ void Unpickler::readGlobal(
                 " has no tensor table\n");
             stack_.emplace_back(tensor_table_->at(data.toInt()));
       });
+    } else if (class_name == "restore_type_tag") {
+      globals_.emplace_back([this] {
+        auto data = stack_.back().toTuple()->elements();
+        stack_.pop_back();
+        TypePtr type = c10::parseType(data.at(1).toStringRef());
```

Comment:
The string ref here is memoized in the serializer, so it would be pretty cheap to have a void* -> TypePtr hashtable in the deserializer that ensures we only parse a type once.